tw-tasks-core/src/main/java/com/transferwise/tasks/helpers/kafka/partitionkey/TaskDefinedPartitionKeyStrategy.java

```diff
@@ -0,0 +1,17 @@
+package com.transferwise.tasks.helpers.kafka.partitionkey;
+
+import com.transferwise.tasks.domain.BaseTask;
+
+public class TaskDefinedPartitionKeyStrategy implements IPartitionKeyStrategy {
+
+  /**
+   * Provides the {@link String} partition key defined in the {@link BaseTask}.
+   *
+   * @param task a Task containing a partition key
+   * @return a possibly null {@link String} to be used as partition key
+   */
+  @Override
+  public String getPartitionKey(BaseTask task) {
+    return task.getPartitionKey();
```

Comment:
Does this really work? I do no see where this is being saved to the DB. Let's imagine leader pod `A` and pod `B`

- B creates the task and assigns the partition key with the method mentioned above. It is written to the [database](https://github.com/transferwise/tw-tasks-executor/blob/master/tw-tasks-core/src/main/java/com/transferwise/tasks/dao/ITaskDao.java#L44)
- A reads the task from the DB. Since it is a basic property it should be null.

You can validate this with an integration test where you set the ID and check if it is still there when the message is sent.

Why do we need that to begin with. Maybe the partition-key should be constructed from the task-ID (our Use-case in the end) then this is not needed. And even if someone wants to set a custom message-key from outside (meaning decoupled from the task) he could utilize the `data` field to set it.

Reply:
oh yeah you're right, it won't. as it is, we are only supporting immediate tasks, or tasks that are resumed manually, but not scheduled tasks. so our use-case is still not covered after all.

> Why do we need that to begin with. Maybe the partition-key should be constructed from the task-ID (our Use-case in the end) then this is not needed. And even if someone wants to set a custom message-key from outside (meaning decoupled from the task) he could utilize the data field to set it.

I don't think using the taskId wouldn't work, because it needs to be unique, and we would need it be repeated to make sure it is processed by the same pod. and data is used to store a payload, so we would need to serialize it again

Reply:
> because it needs to be unique

Why does it need to be unique? In our special case we have this persistence-key as message-key. That can be 
a) either saved as the data-attribute since it is 'just' a sting OR
b) we use this string to generate a UUID. This is a (potential) injective function and guarantees the same UUID based on the same input. Example from fraud that generate the same UUID based on the same transfer-ID: https://github.com/transferwise/fraud-trigger/blob/0be3e106fd89e67eea09fe587c0400c560ee1d4f/service/src/main/java/com/transferwise/fraud/trigger/trigger/TriggerQueue.java#L107

Reply:
> Why does it need to be unique 

Because the taskId is a PrimaryKey on the tasks table

Reply:
Yeah exactly. If persistence-key (and maybe others) are part of the PK no need to create another one on a retry. But this is specific to our use-case. Let's focus here on the lib. And I think we do not need the `TaskDefinedPartitionKeyStrategy`. We have the `RandomPartitionKeyStrategy` which is the current behaviour. A custom implementation can be part of our service and does not need to be in the lib.

Reply:
As we talked, I removed the `TaskDefinedPartitionKeyStrategy` from this PR, and we will implement it on Assembler